### PR TITLE
K-fold datasets in pipeline

### PR DESCRIPTION
**Pull request by jorshi**
Building on @khumairraj's work in #99 to support k-fold datasets

### CODE REVIEW

File: hearpreprocess/pipeline.py [hunk 1]
```diff
@@ -594,9 +622,37 @@ def run(self):
...
+                raise AssertionError(
+                    "Names of splits in metadata don't match the required names for a "
+                    f"presplit_kfold dataset. Expected: {self.task_config['splits']}. "
+                    f"Received: {metadata['split'].unique()}."
+                )
+
+        else:
+            raise ValueError("Unknown split_mode received in task_config")
+
+        assert False
```
Comment (jorshi) [inline hearpreprocess/pipeline.py:655]: Stuck this in to kill the pipeline for testing -- remove it

File: hearpreprocess/pipeline.py [hunk 2]
```diff
@@ -38,16 +40,6 @@
...
 
-# We want no more than 5 hours of audio (training + validation) per task.
-# This can be overriden in the task config.
-# e.g. speech_commands test set.
-# If None, no limit is used.
-MAX_TASK_DURATION_BY_SPLIT = {
-    "train": 3600 * 5 * TRAIN_PERCENTAGE / TRAINVAL_PERCENTAGE,
-    "valid": 3600 * 5 * VALIDATION_PERCENTAGE / TRAINVAL_PERCENTAGE,
-    "test": 3600 * 5 * TEST_PERCENTAGE / TRAINVAL_PERCENTAGE,
-}
```
Comment (jorshi) [inline hearpreprocess/pipeline.py:49]: Getting rid of this - we don't use it, and better to specifically set the duration for each split

File: hearpreprocess/dcase2016_task2.py [hunk 3]
```diff
@@ -72,6 +73,7 @@
                     "md5": "14539d85dec03cb7ac75eb62dd1dd21e",
                 },
             ],
+            "max_task_duration_by_split": {"train": None, "valid": None, "test": None},
```
Comment (turian) [inline hearpreprocess/dcase2016_task2.py:76]: Are we asserting that max_task_duration_by_split must exist for every "mode" if "split_mode" == trainvaltest? Same for split_mode == ...kfold and max_task_duration_by_fold must exist? If not, we should create a separate PR immediately to sanity check this.

Comment (khumairraj) [reply hearpreprocess/dcase2016_task2.py:76]: It is there in the config validator

File: hearpreprocess/pipeline.py [hunk 4]
```diff
@@ -530,6 +522,69 @@ def split_train_test_val(self, metadata: pd.DataFrame):
...
+        """
+        Raises an AssertionError if the set of fold names in task_config doesn't
+        match the set of fold names used for splits in the metadata
+        """
+        if set(metadata["split"].unique()) != set(self.task_config["splits"]):
+            raise AssertionError(
+                "Names of splits in metadata don't match the required names for a "
+                f"kfold dataset. Expected: {self.task_config['splits']}. "
+                f"Received: {metadata['split'].unique()}."
+            )
```
Comment (turian) [inline hearpreprocess/pipeline.py:535]: We're not using splits for kfolds, so why do we need this function?

File: hearpreprocess/pipeline.py [hunk 5]
```diff
@@ -530,6 +522,69 @@ def split_train_test_val(self, metadata: pd.DataFrame):
...
+
+    def split_k_folds(self, metadata: pd.DataFrame):
+        """
+        Deterministically split dataset into k-folds
+        """
+
+        splits_present = metadata["split"].unique()
+        if len(splits_present) > 1:
```
Comment (turian) [inline hearpreprocess/pipeline.py:543]: Again this logic is really confusing. We need some documentation *in one place* of what the expected behavior is for kfold config.py etc. It's really too hard to follow this based upon the code.

File: hearpreprocess/tfds_speech_commands.py [hunk 6]
```diff
@@ -14,6 +14,7 @@
     "version": "0.0.2",
```
Comment (turian) [inline hearpreprocess/tfds_speech_commands.py:14]: Can we make a separate PR that is just all the simple mechanical changes like "trainvaltest" so we can merge that quickly? Not a huge deal, but if it's easy let's do it so we can zoom in on the stuff that is stuff confusing

File: hearpreprocess/runner.py [hunk 7]
```diff
@@ -127,6 +127,28 @@ def run(
...
+
+            if task_config["split_mode"] == "trainvaltest":
+                # Dataset will be partitioned into train/validation/test splits
+                task_config["splits"] = pipeline.SPLITS
+            elif task_config["split_mode"] in ["new_split_kfold", "presplit_kfold"]:
+                # Dataset will be partitioned in k-folds, either using predefined folds
+                # or with using folds defined in the pipeline
+                n_folds = task_config["nfolds"]
+                assert isinstance(n_folds, int)
+                task_config["splits"] = ["fold{:02d}".format(i) for i in range(n_folds)]
```
Comment (turian) [inline hearpreprocess/runner.py:145]: I really don't love this idea of overwriting task_config with undocumented stuff that's not in the schema. How can we avoid that?

File: hearpreprocess/pipeline.py (hunk 4 above)
Comment (khumairraj) [reply hearpreprocess/pipeline.py:535]: `Splits` correspond to the `folds` in the fold mode. So essentially there will always be a split parameter in the task config

File: hearpreprocess/runner.py (hunk 7 above)
Comment (turian) [reply hearpreprocess/runner.py:145]: I mean, can't we make a method here or something? Overwriting the config is really kinda fragile and confusing

File: hearpreprocess/pipeline.py [hunk 8]
```diff
@@ -747,18 +827,7 @@ def run(self):
...
-        if "max_task_duration_by_split" in self.task_config:
-            max_task_duration_by_split = self.task_config["max_task_duration_by_split"]
-            # Check if all the splits are specified in
-            # max_task_duration_by_split
-            assert set(max_task_duration_by_split.keys()) == set(SPLITS)
-            max_split_duration = max_task_duration_by_split[self.split]
-        else:
-            # Get the default values for the split
-            max_split_duration = MAX_TASK_DURATION_BY_SPLIT[self.split]
+        max_split_duration = self.get_max_split_duration()
```
Comment (turian) [inline hearpreprocess/pipeline.py:830]: see like this could have been a dict lookup but now is a function, seems like a nice way to wrap higher-level config stuff.

File: hearpreprocess/pipeline.py [hunk 9]
```diff
@@ -722,6 +777,31 @@ def requires(self):
...
+        if self.task_config["split_mode"] == "trainvaltest":
+            duration_key = "max_task_duration_by_split"
+        else:
+            duration_key = "max_task_duration_by_fold"
+
+        # Assert the the correct max durations are present
+        if duration_key not in self.task_config:
+            raise AssertionError(f"{duration_key} must be defined in task_config")
+
+        max_durations = self.task_config[duration_key]
```
Comment (turian) [inline hearpreprocess/pipeline.py:794]: Wrong behavior for kfold

File: hearpreprocess/runner.py (hunk 7 above)
Comment (khumairraj) [reply hearpreprocess/runner.py:145]: `splits` makes it simple to define the tasks. Like there are some tasks at a split level and the split key helps

File: hearpreprocess/pipeline.py [hunk 10]
```diff
@@ -722,6 +777,31 @@ def requires(self):
             "metadata": self.metadata_task,
         }
 
+    def get_max_split_duration(self):
+        """
+        Returns the max duration for the current split from the task_config
+        """
```
Comment (turian) [inline hearpreprocess/pipeline.py:783]: Needs to be updated. Are we in a place where CI will catch these bugs?

**Comment by turian** [comment]
Generally nice but I'm pushing for a little more clarity, since this can get confusing relatively quickly.

**Review by turian — APPROVED** [review]
